Data Augmentation by Paraphrasing for NLU (#7584)
RasaHQ/rasa: docs/docs/command-line-interface.mdx

|`rasa shell`            |Loads your trained model and lets you talk to your assistant on the command line.                                                         |
|`rasa run`              |Starts a server with your trained model.                                                                                                  |
|`rasa run actions`      |Starts an action server using the Rasa SDK.                                                                                               |
|`rasa visualize`        |Generates a visual representation of your stories.                                                                                        |
|`rasa test`             |Tests a trained Rasa model on any files starting with `test_`.                                                                            |
|`rasa data split nlu`   |Performs a 80/20 split of your NLU training data.                                                                                         |
|`rasa data convert`     |Converts training data between different formats.                                                                                         |

> |`rasa data validate`    |Checks the domain, NLU and conversation data for inconsistencies.                                                                         |

Reviewer: Can you check why these two show different? It could be just a new line char which is fine but just so that the rendered version doesn't look different

Author: Its whitespace changes only according to GitHub - I might have added a space to make then table end pipes align.

> |`rasa data suggest nlu` |Performs data augmentation by paraphrasing and suggests utterances that can be added to the original training data.                       |

Reviewer: ```suggestion
|`rasa data suggest nlu` |Performs data augmentation by suggesting utterances that can be added to the original training data from the augmented dataset.                      |
```

Reviewer: Better to make the change directly in the IDE rather than accepting the suggestion here because of the formatting of the table.

```text [rasa data validate --help]
```

## rasa data suggest nlu

You can perform data augmentation by first [generating paraphrases](https://github.com/RasaHQ/paraphraser) and subsequently running the following command in order to get data suggestions that can be added to the training data:

```bash
rasa data augment nlu
```

> Data augmentation is one strategy that can help produce more training data for intents that currently have little data, are frequently confused, or exhibit low performance in general. Our data augmentation strategy works by generating paraphrases for existing training examples and then evaluating whether adding a subset of these can improve intent classification performance.

Reviewer: ```suggestion
Data augmentation is one strategy that can help produce more training data for intents that currently have little data, are frequently confused, or exhibit low performance in general. Our data augmentation strategy works by feeding the generated paraphrases and then evaluating whether adding a subset of these can improve intent classification performance.
```

Author: Did you mean "feeding the generated paraphrases [to the classifier]" or something similar?

Reviewer: I meant feeding it to the CLI tool because the tool itself doesn't generate paraphrases.

We are evaluating two strategies for selecting paraphrases, firstly by simply randomly choosing a subset of paraphrases, and secondly, by maximising vocabulary expansion, i.e. preferring paraphrases that add new words to a given intent.

The paraphrases should be pre-processed and supplied with similarity scores in order to ensure that noise is minimised (see the [paraphraser repository](https://github.com/RasaHQ/paraphraser) for more details).

:::note
Usage of `rasa data augment nlu` requires a set of generated paraphrases from [https://github.com/RasaHQ/paraphraser](https://github.com/RasaHQ/paraphraser).
:::

:::caution check results
While data augmentation can substantially improve performance for some intents (and therefore also overall NLU performance), it can also degrade the performance of other intents. We therefore recommend you carefully check the generated reports.
:::

You need to specify additional arguments for `rasa data augment nlu`, e.g. the location of your generated paraphrases, your training and evaluation data, as well as an `intent_report.json` from a trained model (without data augmentation) and its config file:

> ```bash

Reviewer: I think we should describe what each argument means in the way how we describe for other commands on this page https://rasa.com/docs/rasa/command-line-interface/#rasa-data-split

Author: Am I correct in assuming that the following within the `mdx` document autogenerates whatever the help command produces? I am not familiar with that bit of markdown magic, though it seems to work for `rasa data split nlu`
```
```text [rasa data augment nlu --help]
`` `
```

rasa data augment nlu --help
```

> 

Reviewer: We should add a line which suggests users to not supply data with entities annotated in them. And if needed annotate the entities for new suggested examples after the augmentation process.

```text [rasa data validate --help]
```

## rasa data suggest nlu

You can perform data augmentation by first [generating paraphrases](https://github.com/RasaHQ/paraphraser) and subsequently running the following command in order to get data suggestions that can be added to the training data:

```bash
rasa data augment nlu
```

This will automatically choose paraphrases and intents to augment from the provided data and train/test two augmented NLU models. You can specify the training data, the evaluation data, the paraphrases, an existing NLU classification report (without augmentation), the model config, and the output directory for storing the augmentation summaries and models using the following arguments:

```text [rasa data augment nlu --help]
```

> 

Author: Err, I thought it should display the same once its rendered. For data split this is in the actual file:

The following arguments are available for `rasa test`:

```text [rasa test --help]
```

As mentioned above, I am not familiar with this bit of markdown magic, though it appears that it triggers the `help` command upon rendering.

```text [rasa data validate --help]
```

> ## rasa data augment nlu

Author: @akelad as discussed on Slack, it would be great if you could review the changes to 
* `docs/docs/command-line-interface.mdx` and
* `docs/docs/generating-nlu-data.mdx`

> ## rasa data augment nlu

Reviewer 2: How about using this one which also points to the forum? https://github.com/RasaHQ/rasa/blob/55a973d617dc5ffc79fb04722e9e43555178c40b/docs/docs/training-data-importers.mdx#L61

```text [rasa data validate --help]
```

## rasa data augment nlu

:::caution

This is an experimental feature.
:::

You can perform data augmentation by first [generating paraphrases](https://github.com/RasaHQ/paraphraser) and subsequently running the following command in order to get data suggestions that can be added to the training data:

```bash
rasa data augment nlu
```

This will automatically choose paraphrases and intents to augment from the provided data and train/test two augmented NLU models. You can specify the training data, the evaluation data, the paraphrases, an existing NLU classification report (without augmentation), the model config, and the output directory for storing the augmentation summaries and models using the following arguments:

```text [rasa data augment nlu --help]
```

> Data augmentation is one strategy that can help produce more training data for intents that currently have little data, are frequently confused, or exhibit low performance in general. Our data augmentation strategy works by feeding the generated paraphrases and then evaluating whether adding a subset of these can improve intent classification performance.

Reviewer 3: is the first part of this paragraph really relevant to this page?

The paraphrases should be pre-processed and supplied with similarity scores in order to ensure that noise is minimised (see the [paraphraser repository](https://github.com/RasaHQ/paraphraser) for more details).

:::note
Usage of `rasa data augment nlu` requires a set of generated paraphrases from [https://github.com/RasaHQ/paraphraser](https://github.com/RasaHQ/paraphraser).

Alternatively, you can use paraphrases from any other resource for data augmentation as long as you make sure they are in the required format as in the example below:

```yaml
nlu:
- intent: ask_query
  examples:
   - text: How can I apply for a new passport?
     metadata:
       paraphrases: |
       - How to apply for passport
       - Apply for a new passport.
       scores: |
       - 0.82
       - 0.93
```

The scores should represent the similarity between the paraphrase and the original sentence. You can calculate similarity scores by e.g. taking the cosine similarity of the encoded sentence vector of the original phrase and the vector respresentations of your paraphrases.

:::

:::note

> Data augmentation does **not** affect entity recognition performance. We furthermore recommend to either remove entity annotations during NLU training or to disable the enitity recognition option in the `DIETClassifier` config.

Reviewer 3: I don't think this should be a note, but just in the regular page

:::note
Usage of `rasa data augment nlu` requires a set of generated paraphrases from [https://github.com/RasaHQ/paraphraser](https://github.com/RasaHQ/paraphraser).

Alternatively, you can use paraphrases from any other resource for data augmentation as long as you make sure they are in the required format as in the example below:

```yaml
nlu:
- intent: ask_query
  examples:
   - text: How can I apply for a new passport?
     metadata:
       paraphrases: |
       - How to apply for passport
       - Apply for a new passport.
       scores: |
       - 0.82
       - 0.93
```

The scores should represent the similarity between the paraphrase and the original sentence. You can calculate similarity scores by e.g. taking the cosine similarity of the encoded sentence vector of the original phrase and the vector respresentations of your paraphrases.

> 

Reviewer: We also need to add a note mentioning that the augmentation strategy does not affect entity performance and it's suggested to either remove annotations for entities during training or turn off entity recognition inside DIET.
